### PR TITLE
Use cached solc version if it is available

### PR DESCRIPTION
**Pull request by eggplantzzz**
During compilation, use a cached version of solc if it one is available that satisfies the version specified in the config.

Issue #1313

### CODE REVIEW

File: packages/truffle-compile/compilerSupplier.js [hunk 1]
```diff
@@ -286,18 +315,32 @@ CompilerSupplier.prototype.getBuilt = function(buildType){
...
 
+/**
+ * Returns a valid version name if compiler file is cached
+ * @param  {String}  localPath
+ * @return {Boolean}
+ */
+CompilerSupplier.prototype.versionIsCached = function(version) {
+  const cachedCompilerFileNames = fs.readdirSync(this.cachePath);
+  const cachedVersions = cachedCompilerFileNames.map((fileName) => {
+    const match = fileName.match(/-v(.*)\+/);
```
Comment (eggplantzzz) [inline packages/truffle-compile/compilerSupplier.js]: Is this a reliable regex to use to parse out version numbers from the compiler names?

File: packages/truffle-compile/compilerSupplier.js [hunk 2]
```diff
@@ -128,18 +131,45 @@ CompilerSupplier.prototype.getDockerTags = function(){
...
+    validVersions[0];
+  return this.getFromCache(compilerFileName);
+}
+
+const getMostRecentVersionOfCompiler = (versions) => {
+  return versions.reduce((mostRecentVersionFileName, fileName) => {
+    const match = fileName.match(/v\d+\.\d+\.\d+.*/);
+    const mostRecentVersionMatch = mostRecentVersionFileName.match(/v\d+\.\d+\.\d+.*/);
+    if (semver.gtr(match[0], mostRecentVersionMatch[0])) return fileName;
+  }, "-v0.0.0+commit");
```
Comment (gnidan) [inline packages/truffle-compile/compilerSupplier.js]: Nice.

File: packages/truffle-compile/test/compilerSupplier.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        supplier.load()
+          .then(() => {
+            assert(supplier.getBuilt.calledWith("docker"));
+            done();
+          });
+      });
+    });
+
+    describe(`when the version is set to a specific word I can't write here`, () => {
+      // Because of how the test script is set up, you can't write native in the strings above and below
```
Comment (gnidan) [inline packages/truffle-compile/test/compilerSupplier.js]: Might be worth a `HACK` comment so it's searchable later

Comment (eggplantzzz) [reply packages/truffle-compile/test/compilerSupplier.js]: Sounds good

**Comment by eggplantzzz** [comment]
One more thing, there are integration tests for the CompilerSupplier in `test_supplier.js`.  I named the file containing the unit specs for the methods related to this PR `compilerSupplier.js` after the file it is testing.  Should we establish some sort of convention for writing unit tests/integration tests and naming the files?

**Comment by gnidan** [comment]
@eggplantzzz I'm in favor of test names matching source names. I will happily encourage you to enforce this sanity :)

**Pull request merged**